### ISSUE TITLE
Show metrics for excluded items in reports

### ISSUE THREAD

**Issue by dlegaultbbry** [gcovr/gcovr#988]
I'd like to show a full report in one go, but have an additional number beside each metric show the stats on excluded items (branch, line, function, etc.).

Something along the lines of `gcovr --no-markers --marker-stats  ...`

This would turn something like
```
lines: 82.6% (161 out of 195)
functions: 100.0% (37 out of 37)
branches: 49.5% (52 out of 105)
conditions: 51.2% (44 out of 86)
```

Into

```
lines: 77.8% (161 out of 207): excl 5.7% (12/207)
functions: 100.0% (37 out of 37): excl 0% (0/37)
branches: 48.6% (52 out of 107): excl 1.8% (2/107)
conditions: 50.0% (44 out of 88): excl 2.2% (2/88)
```

Similar thought for all the other generated formats. 

Not sure if there is any combination of options which can do this today or not. I can work on it and submit a PR if there is nothing like this and it would be a nice addition.

**Comment by Spacetown** [comment]
That sounds good for me but for this we need to extend the data model and change the exclusion handling. For excluded branches we remove them from the data model but for this we need to have an exclusion flag as we do it for lines.
I suggest to do the change in two steps:
1. Add the exclusion flag to the data model and adapt the writers respect it. --> Reference data must not be changed.
2. Add an option (e.g. `--exclusion-stats`) to show the excluded stats. This is not related to markers.

In your  example the percent values are confusing for me because the overall number of elements is inclusive the excluded elements but the percentage is evaluated without it.

**Comment by dlegaultbbry** [comment]
Yes. I wasn't sure if the actual metrics included the lines that were excluded or not. Sounds good on the plan, will work on it.

**Comment by dlegaultbbry** [comment]
I'll be working on this next. Will wait for all the cleanup and linting changes to go in.

**Comment by dlegaultbbry** [comment]
Started working on last part of this feature.

**Comment by dlegaultbbry** [comment]
Had a question regarding the behavior of the option for reporting.

If I understand the stats properly, right now the total for items (branches, etc.) is only for non excluded items.

This means that if I want to report excluded stats, the total now needs to include excluded items in the count.

So something like for branches:
real total = 12
non excluded total = 6
covered = 4

4/6 66.6% -> 4/12 33.3% (excl 6/12 50%)

1) Do you want to change the default reporting to always have total include all items including excluded? 
2) Do you want it to remain the same as now and only inflate the totals with excluded items if --exclusion-stats is given on the command line?

If 1), then this could be done in a separate change. Then I can do the final change to show the exclusion stats as extra on all formats.

**Comment by Spacetown** [comment]
Good question. I would go in direction one with following properties:
- `total`: Overall number of items.
- ´excluded`: Number of excluded items.
- `covered`: Number of covered items.

This should be done with a separate PR which also changes the format version because the meaning of total is changed.

Changing the internal Datta model by using the `--excluded-stats` would result in more problems. What is if you merge trace files which where generated with and without the option?

**Comment by dlegaultbbry** [comment]
Not sure I understand your last question.

I will work on 1) as a separate change.

**Comment by Spacetown** [comment]
This wasn't a question. This was a con for approach 2.

We need to think also if we need the option `--excluded-stats` at all or if we shall always add the excluded counts to the reports. @latk What do you think?

**Comment by latk** [comment]
> if we shall always add the excluded counts to the reports

The problem here is that "excluded" items can have two meanings:

* there is no code here (e.g. excluded via `--exclude-noncode-lines`, `--exclude-unreachable-branches`)
* there is code here, but we don't want to count it for the purpose of calculating coverage statistics (e.g. excluded via marker comments)

This is complicated by things like templates and inline functions, where there is executable source code, but it cannot be counted because it was never compiled to object code. Consider a pseudo-gcov report like this:

```c
    -:  1: inline int not_referenced() {
    -:  2:   printf("code that's invisible\n");
    -:  3:   return 42;
    -:  4: }
    -:  5:
    -:  6: int main(int argc, const char** argv) {
    1:  7:   if (argc != 1) {
#####:  8:     exit(1);  // GCOVR_EXCL_LINE
    -:  9:   }
    1: 10    printf("hello world\n");
    1: 11:   return 0;
    -: 12: }
```

What is the line coverage of that example?

* coverage is 3/3 (I think this is what gcovr currently sees)
* coverage is 3/4 (this is what gcov reports above)
* coverage is 3/5 (counting all code lines that weren't explicitly excluded, including the non-compiled inline function. Probably what most people would expect, but gcovr cannot know this)
* coverage is 3/6 (counting all code lines that could theoretically be executed, but gcovr cannot know this)
* coverage is 3/12 (counting all lines in the file)

The current gcovr data model does not distinguish between different kinds of exclusion, but a hierarchy of line types is explained here:

https://github.com/gcovr/gcovr/blob/e33aeeee2321ffeffaf4b0d70cd024748d05485b/gcovr/coverage.py#L405-L409

As a nested list:

* line
  * excluded
  * reportable
    * covered
    * uncovered

In reports, the "line" and "reportable" concepts tend to be conflated. But "reportable" is also what is expected most of the time.

It can also be difficult to aggregate "excluded" items across multiple files. E.g. if an entire source file is excluded via `--filter/--exclude`, then it won't be processed. This makes percentages that are based on "excluded" items less meaningful. Exclusion handling is done while *reading* coverage data, not while *writing* reports. So exclusion handling affects the internal coverage data model.

All of this is to say: the problem space is super complicated. It is not obvious to me what the best solution is. But given that the concept of "exclusions" is so confusing (even to folks like us who have worked on that code), it could be unhelpful to actively push these metrics onto users. It will confuse 10k people and maybe help 5.

Details might depend on report types:

* `--print-summary` is supposed to be absolutely minimal, adding more info here is probably not helpful for most users
* `--txt` is space-constrained by typical terminal width, adding more info is likely to overload it
* `--json-summary` can be easily extended with additional aggregate metrics and is a good place for surfacing power-user features
* `--html` maybe? Or could users that want this already pull in the data via custom templates?
* `--cobertura` and other existing formats: semantics should stay close to the meaning of those formats, should probably not change metrics to include "excluded" lines.

**Comment by dlegaultbbry** [comment]
Well the reason for the ask is that when auditors/reviewers look for coverage, if the excluded lines are not present in the metrics (at least the summaries), it hides the fact that coverage of the code is incomplete and someone needs to review the reasons why something was excluded (using markers). Dead code is less of an issue as static analyzers can see that separately.

Maybe this can be turned around into a report on the number exclusion markers that are present in source files as an indicator of things are excluded in these files? Maybe add in there a number on the other non marker types of exclusions as a further hint of something might be fishy?

**Comment by Spacetown** [comment]
I was only thinking about the lines and branches excluded by comments. If we do this for all exclusion possibilities this would be quite complicated for the file exclusions. If a file is excluded the file still needs to be processed to get the exclusion metrics.
Can we generate the metrics on line/branch exclusions and ignoring excluded files for this? For each of the following exclusions we should check if they still should be removed or the `excluded` shall be used.
https://github.com/gcovr/gcovr/blob/9ae1533172879a08f0ecb890ddbf06e913f23f45/gcovr/exclusions/__init__.py#L84-L112
In my opinion the `noncode` and `unreachable` should still be removed.

**Comment by dlegaultbbry** [comment]
I'll be honest, I'm not sure what to think about this. I think this should be discussed between you and @latk on a possible way forward. I'll be more than happy to work on it if consensus is reached.

**Comment by latk** [comment]
> I  was only thinking about the lines and branches excluded by comments … In my opinion the `noncode` and `unreachable` should still be removed.

I think that sounds reasonable.

So perhaps the data model needs to be extended to something like this:

* line
  * :sparkles: no data (or noncode, or unreachable)
  * :sparkles: gcov data
    * :sparkles: excluded via marker pattern
    * reportable
      * covered
      * uncovered

I am not sure what the correct hierarchy for this data model is.

Most reports would still only focus on the covered:reportable ratio. On a per-file level, statistics about "excluded via marker pattern" may be meaningful, perhaps this can be surfaced in the HTML-details report. On a global level, file exclusion mechanisms make "excluded via marker pattern" statistics less meaningful, but something like the total count could still be reported in some formats.

**Comment by Spacetown** [comment]
> Most reports would still only focus on the covered:reportable ratio. On a per-file level, statistics about "excluded via marker pattern" may be meaningful, perhaps this can be surfaced in the HTML-details report. On a global level, file exclusion mechanisms make "excluded via marker pattern" statistics less meaningful, but something like the total count could still be reported in some formats.

There is still the question if we shall use a option or we add this info to the report where possible.

**Comment by latk** [comment]
> There is still the question if we shall use a option or we add this info to the report where possible.
> I would would for the second approach.

I agree that avoiding more options is good wherever possible, the CLI is already highly complex.

Whatever reports are adapted, the changes to the data model would have to come first. Making this change separately would also make the changes easy to test, as we would expect that pull request to have zero changes to the reference coverage data in the test suite.

Then, we can think about surfacing this in reports. The JSON formats can clearly accommodate this data, and adding more info to html-details seems possible as well. E.g. highlighting excluded lines, and adding a count of excluded lines in the table in the header.

The txt-summary is a bit more tricky. Aggregate percentages as suggested in the issue at the top are not very meaningful (they can be calculated, but are difficult to interpret). The easiest way might be to add something like `excluded: 123 lines`, without further changes.

**Comment by Spacetown** [comment]
@dlegaultbbry Can you start with the changes to the data model as described in https://github.com/gcovr/gcovr/issues/988#issuecomment-2480737446?

**Comment by dlegaultbbry** [comment]
I will start on it next week. Might take me a while to figure it all out but agree that changing the model will be required and should be done separately. Will that require a bump for both JSON versions?

**Comment by Spacetown** [comment]
Yes, please update the JSON version and the documentation.

**Comment by dlegaultbbry** [comment]
I have not abandoned working on this, just dealing with a few higher priority items at work right now. Will resume working on this early Jan 2025.

**Comment by Spacetown** [comment]
No problem, it's your own request. 😄

**Comment by dlegaultbbry** [comment]
I just didn't want to hold up any new release or anything expecting this to be included! =]

**Comment by Spacetown** [comment]
What do you think about also extending our comment syntax with an option to add a justification for the exclusion?
In that case we need to create a metric with excluded and one with justified.

**Comment by dlegaultbbry** [comment]
Can you give a code example of what it would look like and where would you put the justification in reports? Or only report about number of justified/unjustified exclusions?

**Comment by Spacetown** [comment]
Sorry for my late response. I was thinking about something like:
```C
// GCOVR_EXCL_START | <Reason>: <Free text>
// GCOVR_EXCL_START | EFFORT: Writing a test for this isn't worth the effort
// GCOVR_EXCL_START | HARDWARE: This part depends on the hardware where we do not execute the tests
```

- The Reasons can be a configurable list.
- In the HTML we can add a list of all justifications and link them to the line. We can also work them with a special format.
- In text output or maybe only in the summary we can add a list of found reasons with their count.
- In the excluded metrics we can also distinguish between justified and unjustified numbers.

**Comment by dlegaultbbry** [comment]
> Shall we also add it to the text summary?

It could be added, but I do not know if it's required.

On a related note, do you have a plan for when the next official release will land?

**Comment by Spacetown** [comment]
Maybe this WE. I want to merge #1155 first and need to check something regarding the working directories of GCOV where temporary files are not removed in the tests.
